### ISSUE TITLE
[BUG] UnscheduledInvocationReason: IsPastDue and process does not restart

### ISSUE THREAD

**Issue by jsquire** [Azure/azure-functions-dotnet-worker#534]
@sjuarezgx commented on [Sat Jul 10 2021](https://github.com/Azure/azure-sdk-for-net/issues/22590)

TargetFramework : net5.0
AzureFunctions Version : v3
Function App platform: Linux
Consumption plan
Microsoft.Azure.Functions.Worker.Extensions.Timer : 4.1.0
VS 2019

Hi 
I have a timer trigger function which is scheduled to run once every 10 minutes, and according to the recommendations, the useMonitor property is set to "true".

"bindings": [{
		"schedule": "0 */10 * * * *",
		"useMonitor": true
		"runOnStartup": false,
		"name": "TimerInfo",
		"type": "timerTrigger"
	}]

I have the following error every time the function is triggered (I can reproduce the problem locally also)

[2021-07-10T15:05:17.825Z] Trigger Details: UnscheduledInvocationReason: IsPastDue, OriginalSchedule: 2021-07-10T11:50:00.0000000-03:00
[2021-07-10T15:06:16.410Z] Starting worker process failed
[2021-07-10T15:06:16.413Z] The operation has timed out.
[2021-07-10T15:06:16.452Z] Failed to start a new language worker for runtime: dotnet-isolated.
[2021-07-10T15:06:16.455Z] System.Private.CoreLib: The operation has timed out.

If I set useMonitor = false, then IsPastDue is false, but the ScheduleStatus of my timer is null so I get null reference exception.

I think that this problem is mostly happening on linux platform.

I think that the expected behavior is that the function should run (having useMonitor TRUE ), and in case there is a "past due" it should retry without the worker process to fail.

Besides, I don't know why ScheduleStatus is null when useMonitor = false.

Thank you very much
Sabrina.

**Comment by nbgyxi** [comment]
I experience this as well and I can reproduce it 100% of the time locally. Meaning once my TimerTrigger is PastDue, then it happens 100% of the time. If I remove the TimerTrigger then it happens 0% of the time. If the TimerTrigger is not PastDue then it works fine.

What happens when this error occurs is that it doesn't run any of my code at all, also not any other code than the TimerTrigger.

https://stackoverflow.com/questions/68360871/timer-trigger-with-azure-functions-v3-net-5/68360965

**Comment by OleAlbers** [comment]
I have the same issue, just have to disagree to
> I think that this problem is mostly happening on linux platform.

Fails 100% of the time on Windows when locally debugging in Visual Studio, too

**Comment by chadwillson** [comment]
I am having the same issue on Windows

**Comment by shareonline** [comment]
Is there really not any fix for this? From a developer point of view, the experience (Azure function .NET 5 isolated) is completely broken right now, and im stuck at .NET 3.1 for my functions.. 
Further for some of the latest Graph SDK calls we need the new version 4.0 and this has dependencies into .NET 5.. It's a mess right now, and really high priority in my eyes.

**Comment by shareonline** [comment]
Okay my workaround so far is the following:

#if DEBUG
        const string Timer = "0 */1 * * * *";
#else
        const string Timer = "0 */5 * * * *";
#endif
public async Task RunScheduled([TimerTrigger(Timer, UseMonitor = false)]TimerInfo myTimer, FunctionContext context)

Could also skip UseMonitor and let it run within under 1 minute, but this works for me and can run/debug locally..
whew! Now i can update to .NET 5 with confidense i guess :)

**Comment by click2install** [comment]
It also hangs with a `dotnet-isolated` `TimerTrigger` Function in VS when `RunOnStartup = true`, regardless of the `UseMonitor` value. 

Using `Microsoft.Azure.Functions.Worker.Extensions.Timer` `4.1.0`.

**Comment by ct-haig** [comment]
@shareonline's workaround worked for me for a short while, but it led me to realize something: 

Once the function gets to that `IsPastDue` state, I can reliably get it working again by changing the `TimerTrigger`'s `schedule` argument -- ie, if I change `"0 */5 * * * *"` to `"0 */4 * * * *"`. When it gets stuck again, I just change it back and it begins working again. 

I don't know enough about the functions runtime to know _why_ that helps, but there it is.

**Comment by chadwillson** [comment]
@shareonline thanks so much for the work around. We do need this updated. This is preventing our conversion to .net 5 from moving to production. Please escalate this update. It is un-usable without it.

**Comment by ct-haig** [comment]
@chadwilson - I had assumed this would work in production. Is your experience that this doesn't work when deployed to Azure, either?

**Comment by nbgyxi** [comment]
I have the same issue in development, but it works fine for me in production. I am pretty sure it is an issue related to the tooling on the local development machine.

**Comment by smkanadl** [comment]
> @chadwilson - I had assumed this would work in production. Is your experience that this doesn't work when deployed to Azure, either?

I am pretty sure that it happend to me in production this morning (literally 10 seconds after I red your comment)

**Comment by nbgyxi** [comment]
@smkanadl How do you know? One way is to wait for the next planned invocation to be sure. Or did something get logged about it? Because in dev, nothing gets logged when this issue occurs, that's why I think waiting for the next invocation is probably the only way to be sure.

**Comment by smkanadl** [comment]
@nbgyxi I cannot prove it, but after a deployment, the function was simply not running/stuck and all requests ended up with an nginx 404. I removed the timer, did a new deployment and everything was back to normal.

**Comment by nbgyxi** [comment]
@smkanadl Wow, yes that does sound very much like this issue can occur in production as well. Scary stuff. It could happen at any restart, for example when deploying code.

**Comment by smkanadl** [comment]
@nbgyxi yes, scared me too as I thought it's a "local only" thing. For my current project, I can work without a timer, but for the next project, I'll have to use a timer. I planned to use the isolated worker stuff there too, but this experience let me second guessing...

**Comment by nbgyxi** [comment]
@brettsam I think we need to know if you intend to fix this critical production issue in .NET 5 or whether we should wait for .NET 6 before we dare depend on TimerTriggers?

I think it is fair to categorize it as critical as it can happen on any deployment if the timing is right (wrong) and it causes the entire application to become unresponsive. However, on the other hand, it is pretty rare. You would have to start deployment just before a TimerTrigger is scheduled to run and then when the deployment is over and the app is restarted, then the TimerTrigger would be "IsPastDue" and then we have the issue.

**Comment by chadwilson** [comment]
> @chadwilson - I had assumed this would work in production. Is your experience that this doesn't work when deployed to Azure, either?

FYI, you are looking for chadwillson (2 L's), I am chadwilson (1 L)

**Comment by ct-haig** [comment]
My bad! I appreciate the correction.
________________________________

**Comment by chadwillson** [comment]
I know our timer will fire before a restart would complete, stopping the entire project which has multiple functions. What is the process to get clarity on general timing, so that I can report upline why we have not promoted our conversion. I hope this works in .net 6.

**Comment by extragen** [comment]
When I migrated to .net 5.0 I had registration of `ApplicationInsights` and the timer trigger didn't work for me. The trigger became work when I removed `s.AddApplicationInsightsTelemetry();`.

```
        public static void Main()
        {
            var host = new HostBuilder()
                .ConfigureFunctionsWorkerDefaults()
                .ConfigureServices(s =>
                {
                    s.AddApplicationInsightsTelemetry();
                })
                .Build();
 
            host.Run();
        }
```

**Comment by oyvost** [comment]
Currently a workaround is to start the function app from the cli:
 
func host start --dotnet-isolated-debug

and attach the debugger to the process id.

This is documented here:
https://github.com/Azure/azure-functions-dotnet-worker

**Comment by craigomatic** [comment]
Adding to the list of people who are able to repro - seeing this both locally in VS2019 and deployed to Azure (Windows, isolated consumption function)

**Comment by shareonline** [comment]
> Adding to the list of people who are able to repro - seeing this both locally in VS2019 and deployed to Azure (Windows, isolated consumption function)

Honestly this really needs to be fixed.. But I have to say that my workaround still works okay. Sometimes I hit the timer wrong, but then I just stop the debug, wait for 5 seconds and start it again.. Using it daily since, and for me that works..

**Comment by craigomatic** [comment]
I couldn't get that workaround to work for me so I ended up creating another function app (olden days, in-process style) just for the timer trigger.

**Comment by NoahStahl** [comment]
Also seeing this hang occur when `TimerTrigger` used in a .NET 5 isolated project and latest VS 2019. It's unfortunate that nobody from Microsoft has seen fit to chime in here and communicate anything about it.

In any case, as mentioned in previous comments, I found I could set `UseMonitor = false` in the trigger parameters and  [invoke the function manually through postman](https://docs.microsoft.com/en-us/azure/azure-functions/functions-manually-run-non-http). This at least allows a reasonable debugging experience without the hangs.

**Comment by cbfrank** [comment]
has the same issue when create a new azure functions projects with isolated mode and add time trigger. the VS 2019 template for timetrigger event not work for isolated, after manually add the missing sdk/nuget package, it can resolve the compiler erros, but run timertrigger from VS get VS freezes and azure function worker output timeout error
